pkg/spdk/engine.go

```diff
@@ -957,3 +1153,265 @@ func (e *Engine) SetErrorState() {
+	labels []string, backupTarget string, credential map[string]string, concurrentLimit int32, compressionMethod, storageClassName string, size uint64) (*BackupCreateInfo, error) {
+
+	e.Lock()
+	defer e.Unlock()
+
+	replicaName, replicaAddress := "", ""
+	for name, mode := range e.ReplicaModeMap {
+		if mode != types.ModeRW {
+			continue
+		}
+		replicaName = name
+		replicaAddress = e.ReplicaAddressMap[name]
+		break
+	}
+
+	e.log.Infof("Creating backup %s for volume %s on replica %s address %s", backupName, volumeName, replicaName, replicaAddress)
+
+	replicaServiceCli, err := GetServiceClient(replicaAddress)
+	if err != nil {
+		return nil, grpcstatus.Errorf(grpccodes.Internal, err.Error())
+	}
+
+	recv, err := replicaServiceCli.ReplicaBackupCreate(&client.BackupCreateRequest{
+		BackupName:           backupName,
+		SnapshotName:         snapshotName,
+		VolumeName:           volumeName,
+		ReplicaName:          replicaName,
+		Size:                 size,
+		BackupTarget:         backupTarget,
+		StorageClassName:     storageClassName,
+		BackingImageName:     backingImageName,
+		BackingImageChecksum: backingImageChecksum,
+		CompressionMethod:    compressionMethod,
+		ConcurrentLimit:      concurrentLimit,
+		Labels:               labels,
+		Credential:           credential,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &BackupCreateInfo{
+		BackupName:     recv.Backup,
+		IsIncremental:  recv.IsIncremental,
+		ReplicaAddress: replicaAddress,
+	}, nil
+}
+
+func (e *Engine) BackupStatus(backupName, replicaAddress string) (*spdkrpc.BackupStatusResponse, error) {
+	e.Lock()
+	defer e.Unlock()
+
+	found := false
+	for name, mode := range e.ReplicaModeMap {
+		if e.ReplicaAddressMap[name] == replicaAddress {
+			if mode != types.ModeRW {
+				return nil, grpcstatus.Errorf(grpccodes.Internal, "replica %s is not in RW mode", name)
+			}
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return nil, grpcstatus.Errorf(grpccodes.NotFound, fmt.Sprintf("replica address %s is not found in engine %s for getting backup %v status", replicaAddress, e.Name, backupName))
+	}
+
+	replicaServiceCli, err := GetServiceClient(replicaAddress)
+	if err != nil {
+		return nil, grpcstatus.Errorf(grpccodes.Internal, err.Error())
+	}
+
+	return replicaServiceCli.ReplicaBackupStatus(backupName)
+}
+
+func (e *Engine) getSnapshotsInfo() (map[string]*api.Lvol, error) {
+	lvols := map[string]*api.Lvol{}
+
+	for name, address := range e.ReplicaAddressMap {
+		if e.ReplicaModeMap[name] != types.ModeRW {
+			continue
+		}
+
+		replicaServiceCli, err := GetServiceClient(address)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get service client for replica %s with address %s", name, address)
+		}
+
+		replica, err := replicaServiceCli.ReplicaGet(name)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get replica %s with address %s", name, address)
+		}
```

Comment:
Getting a replica client seems frequently used. 

nit: suggest extra the reused part later.

Reply:
Refactor later.